### ISSUE TITLE
FlutterSerialCOmmunicationPlugin.java shall explicitly register RECEIVER_EXPORTED when registering intent

### ISSUE THREAD

**Issue by ivan2angelov** [farellsujanto/flutter_serial_communication#14]
Unfortunately release v0.2.5 does not solve the issue.
I made Flutter app (it is very much the same as the example app in library code), build it and I get following error:
`Exception has occurred.
PlatformException (PlatformException(error, com.example.flutter_application_1: One of RECEIVER_EXPORTED or RECEIVER_NOT_EXPORTED should be specified when a receiver isn't being registered exclusively for system broadcasts, null, java.lang.SecurityException: com.example.flutter_application_1: One of RECEIVER_EXPORTED or RECEIVER_NOT_EXPORTED should be specified when a receiver isn't being registered exclusively for system broadcasts`

here is the pubspec.yaml I have:
```
name: flutter_application_1
description: "A new Flutter project."
publish_to: 'none'
version: 1.0.0+1

environment:
  sdk: '>=2.19.5 <4.0.0'

dependencies:
  flutter:
    sdk: flutter
  flutter_serial_communication: ^0.2.5
  plugin_platform_interface: '>=2.0.0 <2.1.5'
  cupertino_icons: ^1.0.6

dev_dependencies:
  flutter_test:
    sdk: flutter
  flutter_lints: ^3.0.0

flutter:
  uses-material-design: true
```

When I build the app I made sure that flutter_serial_communication is indeed 0.2.5 -  I checked the dependency file ./app/intermediates/flutter/debug/flutter_build.d and it contains:
```
C:\\Users\\ian\\AppData\\Local\\Pub\\Cache\\hosted\\pub.dev\\flutter_serial_communication-0.2.5\\LICENSE C:\\Users\\ian\\AppData\\Local\\Pub\\Cache\\hosted\\pub.dev\\flutter_serial_communication-0.2.5\\lib\\flutter_serial_communication.dart C:\\Users\\ian\\AppData\\Local\\Pub\\Cache\\hosted\\pub.dev\\flutter_serial_communication-0.2.5\\lib\\flutter_serial_communication_method_channel.dart C:\\Users\\ian\\AppData\\Local\\Pub\\Cache\\hosted\\pub.dev\\flutter_serial_communication-0.2.5\\lib\\flutter_serial_communication_platform_interface.dart C:\\Users\\ian\\AppData\\Local\\Pub\\Cache\\hosted\\pub.dev\\flutter_serial_communication-0.2.5\\lib\\models\\device_info.dart C:\\Users\\ian\\AppData\\Local\\Pub\\Cache\\hosted\\pub.dev\\leak_tracker-10.0.4\\LICENSE
```

What I find interesting is that if I build locally the library downloaded from GitHub it works, but if I use it via pubspec.yaml it fails.

On Android 14 there seems to be a change in handling broadcasting events, see
https://stackoverflow.com/questions/77235063/one-of-receiver-exported-or-receiver-not-exported-should-be-specified-when-a-rec

similar bug reported on https://github.com/react-native-share/react-native-share/issues/1463

I will thus open a new issue.

_Originally posted by @ivan2angelov in https://github.com/farellsujanto/flutter_serial_communication/issues/8#issuecomment-2120217911_

**Comment by farellsujanto** [comment]
Hi, your error is happening when building the app / when connecting into a device?

**Comment by ivan2angelov** [comment]
> connecting

If I download the library and build it locally I see no error.
If I use the library from github and add it via pubspec.yaml I see and exception shown earlier when calling 'FlutterSerialCommunication().connect()` method. The actual line of code causing the Exception is:
https://github.com/farellsujanto/flutter_serial_communication/blob/main/example/lib/main.dart#L54

**Comment by ivan2angelov** [comment]
I think I have a fix. I will submit a branch to demonstrate what (I think) is needed.

**Comment by ivan2angelov** [comment]
I tried, but since I have no write permissions here is the git patch:
```
From b5a4f9e1c3304546ce8eb33e9f212e96beffe33c Mon Sep 17 00:00:00 2001
From: Ivan Angelov <ian@pce-instruments.com>
Date: Mon, 20 May 2024 16:15:30 +0300
Subject: [PATCH] Since Android 14 and Google IO 2023 all Android apps must
 explicitly state if an app export or not export broadcast events.

Link to issue https://github.com/farellsujanto/flutter_serial_communication/issues/14

Link to similal problems discussed on StackOverflow:
https://stackoverflow.com/questions/77235063/one-of-receiver-exported-or-receiver-not-exported-should-be-specified-when-a-rec

Usage of RECEIVER_NOT_EXPORTED requires at least API level 33.
---
 android/build.gradle                          |  2 +-
 .../FlutterSerialCommunicationPlugin.java     | 13 +++-
 example/pubspec.lock                          | 70 +++++++++++++------
 3 files changed, 59 insertions(+), 26 deletions(-)
 mode change 100644 => 100755 example/pubspec.lock

diff --git a/android/build.gradle b/android/build.gradle
index f4d4c8b..c6943f2 100644
--- a/android/build.gradle
+++ b/android/build.gradle
@@ -24,7 +24,7 @@ rootProject.allprojects {
 apply plugin: 'com.android.library'

 android {
-    compileSdkVersion 31
+    compileSdkVersion 33

     compileOptions {
         sourceCompatibility JavaVersion.VERSION_1_8
diff --git a/android/src/main/java/id/farellsujanto/flutter_serial_communication/FlutterSerialCommunicationPlugin.java b/android/src/main/java/id/farellsujanto/flutter_serial_communication/FlutterSerialCommunicationPlugin.java
index a06ea4d..fd9d585 100644
--- a/android/src/main/java/id/farellsujanto/flutter_serial_communication/FlutterSerialCommunicationPlugin.java
+++ b/android/src/main/java/id/farellsujanto/flutter_serial_communication/FlutterSerialCommunicationPlugin.java
@@ -271,9 +271,18 @@ public class FlutterSerialCommunicationPlugin implements FlutterPlugin, MethodCa
       int flags = Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
           ? PendingIntent.FLAG_MUTABLE : 0;
       usbGrantReceiver = new USBGrantReceiver(this);
-      activity.registerReceiver(
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+        activity.registerReceiver(
           usbGrantReceiver,
-          new IntentFilter(PluginConfig.INTENT_ACTION_GRANT_USB));
+          new IntentFilter(PluginConfig.INTENT_ACTION_GRANT_USB),
+          android.content.Context.RECEIVER_NOT_EXPORTED
+          );
+      } else {
+        activity.registerReceiver(
+          usbGrantReceiver,
+          new IntentFilter(PluginConfig.INTENT_ACTION_GRANT_USB)
+          );
+      }
       PendingIntent usbGrantIntent = PendingIntent.getBroadcast(activity,
           0,
           new Intent(PluginConfig.INTENT_ACTION_GRANT_USB), flags);
diff --git a/example/pubspec.lock b/example/pubspec.lock
old mode 100644
new mode 100755
index 1f973b5..f7015ab
--- a/example/pubspec.lock
+++ b/example/pubspec.lock
@@ -37,10 +37,10 @@ packages:
     dependency: transitive
     description:
       name: collection
-      sha256: f092b211a4319e98e5ff58223576de6c2803db36221657b46c82574721240687
+      sha256: ee67cb0715911d28db6bf4af1026078bd6f0128b07a5f66fb2ed94ec6783c09a
       url: "https://pub.dev"
     source: hosted
-    version: "1.17.2"
+    version: "1.18.0"
   cupertino_icons:
     dependency: "direct main"
     description:
@@ -76,12 +76,36 @@ packages:
       path: ".."
       relative: true
     source: path
-    version: "0.2.3"
+    version: "0.2.5"
   flutter_test:
     dependency: "direct dev"
     description: flutter
     source: sdk
     version: "0.0.0"
+  leak_tracker:
+    dependency: transitive
+    description:
+      name: leak_tracker
+      sha256: "7f0df31977cb2c0b88585095d168e689669a2cc9b97c309665e3386f3e9d341a"
+      url: "https://pub.dev"
+    source: hosted
+    version: "10.0.4"
+  leak_tracker_flutter_testing:
+    dependency: transitive
+    description:
+      name: leak_tracker_flutter_testing
+      sha256: "06e98f569d004c1315b991ded39924b21af84cf14cc94791b8aea337d25b57f8"
+      url: "https://pub.dev"
+    source: hosted
+    version: "3.0.3"
+  leak_tracker_testing:
+    dependency: transitive
+    description:
+      name: leak_tracker_testing
+      sha256: "6ba465d5d76e67ddf503e1161d1f4a6bc42306f9d66ca1e8f079a47290fb06d3"
+      url: "https://pub.dev"
+    source: hosted
+    version: "3.0.1"
   lints:
     dependency: transitive
     description:
@@ -94,34 +118,34 @@ packages:
     dependency: transitive
     description:
       name: matcher
-      sha256: "1803e76e6653768d64ed8ff2e1e67bea3ad4b923eb5c56a295c3e634bad5960e"
+      sha256: d2323aa2060500f906aa31a895b4030b6da3ebdcc5619d14ce1aada65cd161cb
       url: "https://pub.dev"
     source: hosted
-    version: "0.12.16"
+    version: "0.12.16+1"
   material_color_utilities:
     dependency: transitive
     description:
       name: material_color_utilities
-      sha256: "9528f2f296073ff54cb9fee677df673ace1218163c3bc7628093e7eed5203d41"
+      sha256: "0e0a020085b65b6083975e499759762399b4475f766c21668c4ecca34ea74e5a"
       url: "https://pub.dev"
     source: hosted
-    version: "0.5.0"
+    version: "0.8.0"
   meta:
     dependency: transitive
     description:
       name: meta
-      sha256: "3c74dbf8763d36539f114c799d8a2d87343b5067e9d796ca22b5eb8437090ee3"
+      sha256: "7687075e408b093f36e6bbf6c91878cc0d4cd10f409506f7bc996f68220b9136"
       url: "https://pub.dev"
     source: hosted
-    version: "1.9.1"
+    version: "1.12.0"
   path:
     dependency: transitive
     description:
       name: path
-      sha256: "8829d8a55c13fc0e37127c29fedf290c102f4e40ae94ada574091fe0ff96c917"
+      sha256: "087ce49c3f0dc39180befefc60fdb4acd8f8620e5682fe2476afd0b3688bb4af"
       url: "https://pub.dev"
     source: hosted
-    version: "1.8.3"
+    version: "1.9.0"
   plugin_platform_interface:
     dependency: transitive
     description:
@@ -147,18 +171,18 @@ packages:
     dependency: transitive
     description:
       name: stack_trace
-      sha256: c3c7d8edb15bee7f0f74debd4b9c5f3c2ea86766fe4178eb2a18eb30a0bdaed5
+      sha256: "73713990125a6d93122541237550ee3352a2d84baad52d375a4cad2eb9b7ce0b"
       url: "https://pub.dev"
     source: hosted
-    version: "1.11.0"
+    version: "1.11.1"
   stream_channel:
     dependency: transitive
     description:
       name: stream_channel
-      sha256: "83615bee9045c1d322bbbd1ba209b7a749c2cbcdcb3fdd1df8eb488b3279c1c8"
+      sha256: ba2aa5d8cc609d96bbb2899c28934f9e1af5cddbd60a827822ea467161eb54e7
       url: "https://pub.dev"
     source: hosted
-    version: "2.1.1"
+    version: "2.1.2"
   string_scanner:
     dependency: transitive
     description:
@@ -179,10 +203,10 @@ packages:
     dependency: transitive
     description:
       name: test_api
-      sha256: "75760ffd7786fffdfb9597c35c5b27eaeec82be8edfb6d71d32651128ed7aab8"
+      sha256: "9955ae474176f7ac8ee4e989dadfb411a58c30415bcfb648fa04b2b8a03afa7f"
       url: "https://pub.dev"
     source: hosted
-    version: "0.6.0"
+    version: "0.7.0"
   vector_math:
     dependency: transitive
     description:
@@ -191,14 +215,14 @@ packages:
       url: "https://pub.dev"
     source: hosted
     version: "2.1.4"
-  web:
+  vm_service:
     dependency: transitive
     description:
-      name: web
-      sha256: dc8ccd225a2005c1be616fe02951e2e342092edf968cf0844220383757ef8f10
+      name: vm_service
+      sha256: "3923c89304b715fb1eb6423f017651664a03bf5f4b29983627c4da791f74a4ec"
       url: "https://pub.dev"
     source: hosted
-    version: "0.1.4-beta"
+    version: "14.2.1"
 sdks:
-  dart: ">=3.1.0-185.0.dev <4.0.0"
-  flutter: ">=2.5.0"
+  dart: ">=3.3.0 <4.0.0"
+  flutter: ">=3.18.0-18.0.pre.54"
--
2.34.1
```

**Comment by farellsujanto** [comment]
you can create a new branch then push your changes & create PR @ivan2angelov

**Comment by JoaquinGarciaFivecomm** [comment]
Hi,

I’m running into the same problems with the FlutterSerialCommunication library on Android 14/API 34 due to the new rules for USB or serial connections. Google Play is also pushing for an API level update by August 31, 2024, to keep my app updated on the store.

Is there an update in the works to handle these Android changes? Any info would help!

Thanks!

**Comment by farellsujanto** [comment]
Oh I forgot to implement this, thanks for bumping this @JoaquinGarciaFivecomm , will update the library around mid next week

**Comment by AlfaizTenup** [comment]
@farellsujanto 
Any update on this.
we are stuck for android 14 release.

**Comment by farellsujanto** [comment]
I have released v0.2.6, please do update to that version @AlfaizTenup
